hydromt/components/datasets.py

```diff
@@ -0,0 +1,407 @@
+        self._filename = filename
+        self._tmp_data_dir = None
+        self._defered_file_closes = []
+        super().__init__(model=model)
+
+    @property
+    def data(self) -> XArrayDict:
+        """Model data in the form of xarray objects.
+
+        Return dict of xarray.Dataset or xarray.DataArray objects
+        """
+        if self._data is None:
+            self._initialize()
+
+        assert self._data is not None
+        return self._data
+
+    def _initialize(self, skip_read=False) -> None:
+        """Initialize geoms."""
+        if self._data is None:
+            self._data = dict()
+            if self._root.is_reading_mode() and not skip_read:
+                self.read()
+
+    def set(
+        self,
+        data: Union[Dataset, DataArray],
+        name: Optional[str] = None,
+        split_dataset: bool = False,
+    ):
+        """Add data to the xrarray component.
+
+        Arguments
+        ---------
+        data: xarray.Dataset or xarray.DataArray
+            New xarray object to add
+        name: str
+            name of the xarray.
+        """
+        self._initialize()
+        if isinstance(data, DataFrame):
+            data = data.to_xarray()
+        assert self._data is not None
+        if split_dataset:
+            if isinstance(data, Dataset):
+                ds = {str(name): data[name] for name in data.data_vars}
+            else:
+                raise ValueError(
+                    f"Can only split dataset for Datasets not for {type(data).__name__}"
+                )
+        elif name is not None:
+            ds = {name: data}
+        elif data.name is not None:
+            ds = {str(data.name): data}
+        else:
+            raise ValueError(
+                "Either name should be set, the data needs to have a name or split_dataset should be True"
+            )
+
+        ds = cast(XArrayDict, ds)
+
+        for name, d in ds.items():
+            if name in self._data:
+                self._logger.warning(f"Replacing xarray: {name}")
+            self._data[name] = d
+
+    @hydromt_step
+    def read(
+        self, filename: Optional[str] = None, single_var_as_array: bool = True, **kwargs
+    ) -> None:
+        r"""Read model dataset files at <root>/<filename>.
+
+        key-word arguments are passed to :py:func:`hydromt.io.readers.read_nc`
+
+        Parameters
+        ----------
+        filename : str, optional
+            filename relative to model root. should contain a {name} placeholder
+            which will be used to determine the names/keys of the geometries.
+            if None, the path that was provided at init will be used.
+        **kwargs:
+            Additional keyword arguments that are passed to the
+            `hydromt.io.readers.read_nc` function.
+        """
+        self._root._assert_read_mode()
+        self._initialize(skip_read=True)
+        kwargs = {**{"engine": "netcdf4"}, **kwargs}
```

Comment:
All the conversation got outdated but I like your explanation :) Maybe so that user can overwrite, can you instead check if engine not in kwargs then engine =netcdf4 or does this line takes care of that somehow? That the original kwargs get priority?

Reply:
This does the same, if `engine` is in `kwargs` it will get priority. Though if you like I'm happy to do it the manual way.